### PR TITLE
Include full error message details in exception messages.

### LINKED ISSUE #811: Use specific exception classes where appropriate
## Description

I tried to read a file that doesn't exist with `sim.create_state_from_gsd`. Should GSD file reading errors be `OSError` and/or `FileNotFoundError` instead of `RuntimeError`?

```pytb
**ERROR**: dump.gsd: Bad file descriptor - lattice.gsd

---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-3-3623b04ad812> in <module>
     11 
     12 sim.operations.integrator = mc
---> 13 sim.create_state_from_gsd(filename='lattice.gsd')

~/miniconda3/envs/dice/lib/python3.8/site-packages/hoomd/simulation.py in create_state_from_gsd(self, filename, frame)
    107                                         self.device._cpp_exec_conf)
    108         # Grab snapshot and timestep
--> 109         reader = _hoomd.GSDReader(self.device._cpp_exec_conf,
    110                                   filename, abs(frame), frame < 0)
    111         snapshot = Snapshot._from_cpp_snapshot(reader.getSnapshot(),

RuntimeError: Error reading GSD file
```

## Proposed solution

Use another exception instead of `RuntimeError`, perhaps `FileNotFoundError` or `OSError`.

## Additional context

I think this may apply to other exceptions as well - we use `RuntimeError` pretty liberally in our code bases and it seems like it's a catch-all for things that have more specific causes. The [docstring](https://docs.python.org/3/library/exceptions.html#RuntimeError) says
> Raised when an error is detected that doesn’t fall in any of the other categories. The associated value is a string indicating what precisely went wrong.

I frequently refer to this list of all built-in exceptions when choosing what to raise. It may be useful to other developers. https://docs.python.org/3/library/exceptions.html#exception-hierarchy

### PR DESCRIPTION
**Pull request by joaander**
## Description

Report full error message in the exception messages. Remove them from the `m_exec_conf->error()` output.

## Motivation and context

1) In some cases, the exception would tear down stderr before the full error message would be printed.
2) Even if the error message was printed, this change gives the user the full details of the error in one place rather than spread across many lines of output.

Resolves #811 

## How has this been tested?

The changes compile.

## Change log

```
Changed:

* Report full error message in the exception messages.
```

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/master/ContributorAgreement.md).
- [x] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.

### CODE REVIEW

File: hoomd/BondedGroupData.cc [hunk 1]
```diff
@@ -490,9 +490,10 @@ BondedGroupData<group_size, Group, name, has_type_mapping>::getGroupByTag(unsign
         {
         if (group_idx == GROUP_NOT_LOCAL)
             {
-            m_exec_conf->msg->error() << "Trying to get type or constraint value of " << name << " "
-                                      << tag << " which does not exist!" << endl;
-            throw runtime_error(std::string("Error getting ") + name);
+            std::ostringstream s;
+            s << "Trying to get type or constraint value of " << name << " " << tag
+              << " which does not exist!";
+            throw runtime_error(std::string(s.str()));
```
Comment (b-butler) [inline hoomd/BondedGroupData.cc:496]:
```suggestion
            throw runtime_error(s.str());
```

File: hoomd/md/NeighborList.cc [hunk 2]
```diff
@@ -1103,20 +1101,16 @@ void NeighborList::addOneThreeExclusionsFromTopology()
 
         if (nBondsA >= MAXNBONDS)
             {
-            m_exec_conf->msg->error()
-                << "nlist: Too many bonds to process exclusions for particle with tag: " << tagA
-                << endl;
-            m_exec_conf->msg->error() << "Maximum allowed is currently: " << MAXNBONDS - 1 << endl;
```
Comment (b-butler) [inline hoomd/md/NeighborList.cc:1109]: Do we not want to inform users of the maximum bonds they can have?

File: hoomd/md/PPPMForceCompute.cc [hunk 3]
```diff
@@ -91,38 +89,31 @@ void PPPMForceCompute::setParams(unsigned int nx,
 
         if (!is_pow2(m_mesh_points.x) || !is_pow2(m_mesh_points.y) || !is_pow2(m_mesh_points.z))
             {
-            m_exec_conf->msg->error()
-                << "The number of mesh points along the every direction must be a power of two!"
-                << std::endl;
-            throw std::runtime_error("Error initializing charge.pppm");
+            std::ostringstream s;
+            s << "The number of mesh points along the every direction must be a power of two!";
+            throw std::runtime_error(s.str());
```
Comment (b-butler) [inline hoomd/md/PPPMForceCompute.cc:94]:
```suggestion
            throw std::runtime_error(
                "The number of mesh points along the every direction must be a power of two!");
```

**Review by b-butler — CHANGES_REQUESTED** [review]
I only have a few small comments.

File: hoomd/md/NeighborList.cc (hunk 2 above)
Comment (joaander) [reply hoomd/md/NeighborList.cc:1109]: Done.

**Review by b-butler — COMMENTED** [review]
:+1:

**Review by b-butler — APPROVED** [review]

**Review by mariaward10 — COMMENTED** [review]
It looks good to go!

**Pull request merged**